python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+
+    def wait(self, timeout=None):
+        start = time.time()
+        while timeout is None or time.time() - start < timeout:
+            if self._result_thread.done():
+                break
+            time.sleep(0.001)
+
+    def get(self, timeout=None):
+        self.wait(timeout)
+        if not self._result_thread.done():
+            raise TimeoutError
+
+        results = []
+        for batch in self._result_thread.results():
+            for result in batch:
+                if isinstance(result, PoolTaskError):
+                    raise result.underlying
+            results.extend(batch)
+        if self._single_result:
+            return results[0]
+
+        return results
+
+    def ready(self):
+        return self._result_thread.done()
+
+    def successful(self):
+        if not self.ready():
+            raise ValueError("{0!r} not ready".format(self))
+        return not self._result_thread.got_error()
+
+
+class IMapIterator(object):
+    def __init__(self, pool, func, iterable, chunksize=None):
+        self._pool = pool
+        self._func = func
+        self._next_chunk_index = 0
+        self._chunk_object_ids = []
+        self._ready_indices = []
+        self._ready_objects = []
+        if not hasattr(iterable, "__len__"):
+            iterable = [iterable]
+        self._iterator = iter(iterable)
+        if chunksize:
+            self._chunksize = chunksize
+        else:
+            self._chunksize = pool._calculate_chunksize(iterable)
+        self._total_chunks = int(math.ceil(len(iterable) / chunksize))
+        self._result_thread = ResultThread(
+            [], total_object_ids=self._total_chunks)
+        self._result_thread.start()
+
+        for _ in range(len(pool._actor_pool)):
+            self._submit_next_chunk()
+
+    def _submit_next_chunk(self):
+        if len(self._chunk_object_ids) >= self._total_chunks:
+            return
+
+        actor_index = len(self._chunk_object_ids) % len(self._pool._actor_pool)
+        new_chunk_id = self._pool._submit_chunk(self._func, self._iterator,
+                                                self._chunksize, actor_index)
+        self._chunk_object_ids.append(new_chunk_id)
+        self._ready_indices.append(None)
+        self._result_thread.add_object_id(new_chunk_id)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.next()
+
+
+class OrderedIMapIterator(IMapIterator):
+    def next(self, timeout=None):
+        if len(self._ready_objects) != 0:
+            return self._ready_objects.pop(0)
+
+        if self._next_chunk_index == len(self._chunk_object_ids):
+            raise StopIteration
+
+        while timeout is None or timeout > 0:
+            start = time.time()
+            try:
+                index = self._result_thread.next_ready_index(timeout=timeout)
+                self._submit_next_chunk()
+            except queue.Empty:
+                raise TimeoutError
```

Comment:
This is confusing, why is it a timeout error if the queue is empty?

Reply:
raising `Empty` is how `queue.Queue` signals a timeout. Leaving comment about it.

Reply:
Ohh. Huh, I wonder if that's confusing. Going from `TimeoutError` to `queue.Empty` makes sense when you're dealing with a queue, since it means the queue was found to be empty during that entire time. But going in the reverse direction seems surprising, since the queue being empty doesn't always imply the timeout has actually elapsed (you might've asked for a non-blocking call instead).

It seems in this case you only actually do blocking calls, so I guess that aspect is fine—but in that case, would it be possible to translate the exception at the `Queue.get` call site instead? It's not obvious what's going on here. (And also if the code changes later to have blocking calls, it'll actually be confusing.)

Reply:
Sure that's better